test/SILOptimizer/devirtualize_protocol_composition.swift

```diff
@@ -0,0 +1,138 @@
+// RUN: %target-swift-frontend -parse-as-library -O -wmo -emit-sil  %s | %FileCheck %s
+// RUN: %target-swift-frontend -parse-as-library -Osize -wmo -emit-sil  %s | %FileCheck %s
+
+// This is an end-to-end test to ensure that the optimizer devertualizes
+// calls to a protocol composition type.
+
+public class ClassA<T> { }
+
+protocol ProtocolA {
+  func foo() -> Int
+}
+
+protocol ProtocolB {
+  func bar() -> Int
+}
+
+public class ClassB: ClassA<String> {
+  func foo() -> Int {
+    return 10
+  }
+}
+
+extension ClassB: ProtocolA { }
+
+public class ClassC<T>: ClassA<T> {
+  func foo() -> Int {
+    return 10
+  }
+}
+
+extension ClassC: ProtocolA { }
+
+public class ClassD { }
+public class ClassE : ClassD {
+  func foo() -> Int {
+    return 10
+  }
+}
+
+extension ClassE: ProtocolA { }
+
+public class ClassF {
+  func foo() -> Int {
+    return 10
+  }
+  
+  func bar() -> Int {
+    return 10
+  }
+}
+
+extension ClassF: ProtocolA, ProtocolB { }
+
+public class ClassG <T> {
+  func foo() -> Int {
+    return 10
+  }
+  
+  func bar() -> Int {
+    return 10
+  }
+}
+
+extension ClassG: ProtocolA, ProtocolB { }
+
+public class ClassH {
+  typealias type = ClassD
+}
+
+func shouldOptimize1<T>(_ x: ClassA<T> & ProtocolA) -> Int {
+  return x.foo()
+}
+
+func shouldOptimize2(_ x: ClassD & ProtocolA) -> Int {
+  return x.foo()
+}
+
+func shouldOptimize3(_ x: ProtocolA & ProtocolB) -> Int {
+  return x.foo() + x.bar()
+}
+
+func shouldOptimize4(_ x: ProtocolA & ProtocolB) -> Int {
+  return x.foo() + x.bar()
+}
+
+func shouldOptimize5(_ x: ClassH.type & ProtocolA) -> Int {
+  return x.foo()
+}
+
+//CHECK: entryPoint1
+//CHECK-NOT: init_existential_ref
+//CHECK-NOT: open_existential_ref
+//CHECK-NOT: witness_method
+//CHECK: return
+public func entryPoint1(c: ClassB) -> Int {
+  return shouldOptimize1(c)
+}
+
+// TODO: create SR -- this causes a crash on master too
```

Comment:
[Here's an example.](https://swift.godbolt.org/z/FOPey2) This crashes because of the issue #28012 resolved but even after that commit is applied, it still crashes because of another issue.

Reply:
Great catch. Now that you've been looking at this code, can you see what the problem is?

Reply:
I'll take another look and see if I can figure it out.